### ISSUE TITLE
Refactor `Storage` trait

### ISSUE THREAD

**Issue by spetz** [apache/iggy#183]
For example, the `load()` method expects an empty struct to be passed to load the data, while it's not the case anymore for some components.

**Comment by YHM404** [comment]
Seems that refactor to `async fn load(&self) -> Result<T, Error>;` would be better?

**Comment by YHM404** [comment]
> Seems that refactor to `async fn load(&self) -> Result<T, Error>;` would be better?

If you guys think it's a good idea, I'll go ahead and refactor it.

**Comment by YHM404** [comment]
I have the following suggestions for refactoring the Storage trait to make it more intuitive and extensible

**Comment by spetz** [comment]
Yes, this sounds good, ideally, the storage would just return the data and do not expect mutable struct passed into it to change its internal state. Most of the new methods already work in this way, but the older ones left intact.

**Comment by YHM404** [comment]
> Yes, this sounds good, ideally, the storage would just return the data and do not expect mutable struct passed into it to change its internal state. Most of the new methods already work in this way, but the older ones left intact.

I have the following suggestions for refactoring the Storage trait to make it more intuitive and extensible
```
pub trait WithKey {
    type Key;
    fn get_key(&self) -> &Self::Key;
}

#[async_trait]
pub trait NewStorage<T, K>: Sync + Send
where
    T: WithKey<Key = K>,
{
    async fn load(&self, key: &K) -> Result<(), Error>;
    async fn save(&self, component: T) -> Result<(), Error>;
    async fn delete(&self, key: &K) -> Result<(), Error>;
}
```
but it's a bit difficult to refactor, do you have any suggestions for step-by-step refactoring?
